certora/specs/DoubleEndedQueue.spec

```diff
@@ -0,0 +1,408 @@
+import "helpers.spec"
+
+methods {
+    pushFront(bytes32)                    envfree
+    pushBack(bytes32)                     envfree
+    popFront()          returns (bytes32) envfree
+    popBack()           returns (bytes32) envfree
+    clear()                               envfree
+
+    // exposed for FV
+    begin()             returns (int128)  envfree
+    end()               returns (int128)  envfree
+    
+    // view
+    length()            returns (uint256) envfree
+    empty()             returns (bool)    envfree
+    front()             returns (bytes32) envfree
+    back()              returns (bytes32) envfree
+    at_(uint256)        returns (bytes32) envfree // at is a reserved word
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Helpers                                                                                                             │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+
+function min_int128() returns mathint {
+    return (2 ^ 128) / 2 * -1;
```

Comment:
Also, is there any reason you use `^` and not `<<`? In some languages `^` is power, but in other its xor (and `**` is power). To avoid the confusion i's use shifts.

Reply:
Added the <<, I think it's now clearer. I decided not doing it at first just to play it safe